### PR TITLE
vmagent: remote write respect Retry-After in header

### PR DESCRIPTION
**Pull request by jiekun**
### Describe Your Changes
related issue: https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6097

#### Changed
- Remote write retry policy in `vmagent` is changed into:
  1. Respect `Retry-After` duration if exists.
  2. Otherwise, calculate next retry duration by backoff policy (x2) and max retry duration limit.
 
#### Docs
- `CHANGELOG.md`.

---
### Checklist
The following checks are mandatory:

- [x] My change adheres [VictoriaMetrics contributing guidelines](https://docs.victoriametrics.com/contributing/).

### CODE REVIEW

**Review by hagen1778 — CHANGES_REQUESTED** [review]

File: app/vmagent/remotewrite/client.go [hunk 1]
```diff
@@ -474,3 +474,42 @@ again:
...
+	}
+
+	if retryAfterString == "" {
+		return retryDuration
+	}
+
+	var retryAfterDuration time.Duration
+
+	// Retry-After could be in "Mon, 02 Jan 2006 15:04:05 GMT" format.
+	parsedTime, err := time.Parse(http.TimeFormat, retryAfterString)
```
[the remark this answers is not in the document]
Comment (jiekun) [reply app/vmagent/remotewrite/client.go:516]: It's now seperated from `calculateRetryDuration` and called in `sendBlockHTTP`

File: app/vmagent/remotewrite/client_test.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	"net/http"
+	"testing"
+	"time"
+)
+
+func TestCalculateRetryDuration(t *testing.T) {
+	// testFunc evaluate if the result of calculateRetryDuration is
+	// 1. >= expectMinDuration
+	// 2. <= expectMinDuration + 10% (see timeutil.AddJitterToDuration)
+	testFunc := func(name string, retryAfterString string, retryDuration, expectMinDuration time.Duration) {
```
[the remark this answers is not in the document]
Comment (jiekun) [reply app/vmagent/remotewrite/client_test.go]: I love this idea, and it brings up some scenarios that I had already forgotten.

Please review the test cases in the `Call calculateRetryDuration for multiple times` block.

File: app/vmagent/remotewrite/client.go [hunk 3]
```diff
@@ -474,3 +474,42 @@ again:
 }
 
 var remoteWriteRejectedLogger = logger.WithThrottler("remoteWriteRejected", 5*time.Second)
+
+// calculateRetryAfterDuration calculate the retry duration.
+// 1. Calculate next retry duration by backoff policy (x2) and max retry duration limit.
+// 2. If Retry-After exists in response header, use max(Retry-After duration, next retry duration).
```
[the remark this answers is not in the document]
Comment (jiekun) [reply app/vmagent/remotewrite/client.go]: I have simplified it and removed unrelated content.

File: app/vmagent/remotewrite/client.go [hunk 4]
```diff
@@ -481,3 +480,48 @@ again:
...
+	if err == nil {
+		retryAfterDuration = time.Duration(time.Until(parsedTime).Seconds()) * time.Second
+	} else {
+		// Retry-After could be in seconds.
+		if d, err := strconv.Atoi(retryAfterString); err == nil {
+			retryAfterDuration = time.Duration(d) * time.Second
+		}
+	}
+
+	return retryAfterDuration
```
Comment (zekker6) [inline app/vmagent/remotewrite/client.go:526]:
```suggestion
	if parsedTime, err := time.Parse(http.TimeFormat, retryAfterString); err == nil {
		return time.Duration(time.Until(parsedTime).Seconds()) * time.Second
	}
	// Retry-After could be in seconds.
	if d, err := strconv.Atoi(retryAfterString); err == nil {
		return time.Duration(d) * time.Second
	}

	return 0
```

Comment (jiekun) [reply app/vmagent/remotewrite/client.go]: Thank you! It has been simplified accordingly.

**Review by zekker6 — APPROVED** [review]

File: app/vmagent/remotewrite/client.go [hunk 5]
```diff
@@ -481,3 +480,46 @@ again:
 }
 
 var remoteWriteRejectedLogger = logger.WithThrottler("remoteWriteRejected", 5*time.Second)
+
+// calculateRetryAfterDuration calculate the retry duration.
+// 1. Calculate next retry duration by backoff policy (x2) and max retry duration limit.
+// 2. use max(Retry-After duration, next retry duration).
+//
+// Also see: https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6097
+func calculateRetryDuration(retryAfterDuration, retryDuration, maxRetryDuration time.Duration) time.Duration {
```
Comment (hagen1778) [inline app/vmagent/remotewrite/client.go:489]: Hey @jiekun @zekker6 

If server responds with exact Retry-After value, shouldn't we respect it as is? I mean, shouldn't we completely ignore default backoff policy? It seems like what users would expect to happen if they gone to extent of sending the specific Retry-After value

File: app/vmagent/remotewrite/client_test.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+	"testing"
+	"time"
+)
+
+func TestCalculateRetryDuration(t *testing.T) {
+	// `testFunc` call `calculateRetryDuration` for `n` times
+	// and evaluate if the result of `calculateRetryDuration` is
+	// 1. >= expectMinDuration
+	// 2. <= expectMinDuration + 10% (see timeutil.AddJitterToDuration)
+	testFunc := func(name string, retryAfterDuration, retryDuration time.Duration, n int, expectMinDuration time.Duration) {
```
Comment (hagen1778) [inline app/vmagent/remotewrite/client_test.go:15]: Would you mind updating test cases with `f-tests` approach like in rest of tests in VM?
Please see https://itnext.io/f-tests-as-a-replacement-for-table-driven-tests-in-go-8814a8b19e9e

File: app/vmagent/remotewrite/client.go [hunk 7]
```diff
@@ -456,10 +456,9 @@ again:
 
 	// Unexpected status code returned
 	retriesCount++
-	retryDuration *= 2
-	if retryDuration > maxRetryDuration {
-		retryDuration = maxRetryDuration
-	}
+	retryDuration = calculateRetryDuration(parseRetryAfterHeader(resp.Header.Get("Retry-After")), retryDuration, maxRetryDuration)
```
Comment (hagen1778) [inline app/vmagent/remotewrite/client.go:459]: I think we should log whether we got Retry-After header, and log if we got it but failed to parse.

**Review by hagen1778 — CHANGES_REQUESTED** [review]

File: app/vmagent/remotewrite/client.go (hunk 7 above)
Comment (jiekun) [reply app/vmagent/remotewrite/client.go:466]: How about logging them both with 1 line:
```
2024-09-04T03:34:02.237Z	info	app/vmagent/remotewrite/client.go:520	found Retry-After header 10, parsed into 10.00 second(s)
2024-09-04T03:34:02.237Z	info	app/vmagent/remotewrite/client.go:520	found Retry-After header Wed, 04 Sep 2024 03:34:32 GMT, parsed into 29.00 second(s)
2024-09-04T03:34:02.237Z	info	app/vmagent/remotewrite/client.go:520	found Retry-After header invalid-retry-after, parsed into 0.00 second(s)
2024-09-04T03:34:02.237Z	info	app/vmagent/remotewrite/client.go:520	found Retry-After header Wed, 04 Sep 2024 11:34:12 FAKETZ, parsed into 0.00 second(s)
```

It should be simple enough for user to know whether it's parsed correctly.

File: app/vmagent/remotewrite/client.go (hunk 5 above)
Comment (jiekun) [reply app/vmagent/remotewrite/client.go:496]: Updated as mentioned above. I'd like to discuss: 
- Should we also respect the limit of `--remoteWrite.retryMinInterval` and `--remoteWrite.retryMaxTime`? So:
    1. If `RA` header exist, retry == `max(min(RA, retryMaxTime), retryMinInterval)`;
    2. If `RA` header not exist, use default back-off.

File: app/vmagent/remotewrite/client_test.go (hunk 6 above)
Comment (jiekun) [reply app/vmagent/remotewrite/client_test.go]: done.

File: app/vmagent/remotewrite/client.go (hunk 5 above)
Comment (hagen1778) [reply app/vmagent/remotewrite/client.go]: I think if RA exists, client should respect it.
If not, fall back to the default behavior.

**Review by hagen1778 — APPROVED** [review]
Thanks for the contribution!

**Pull request merged**